<a href="https://colab.research.google.com/github/doyinsolamiolaoye/hamoye_internship/blob/master/hamoye_stage_two_quiz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 # Machine Learning: Regression - Predicting Energy Efficiency of Buildings




## Introduction

According to the United Nations Environmental Program (UNEP) Sustainable Buildings and Climate Initiative, construction trade contributes as much as 30% to all global greenhouse gas emissions and consumes up to 40% of all energy used worldwide. Climate change is currently having a powerful impact on how buildings are designed and constructed. In this Project, I will develop a multivariate multiple regression model to study this effect.

## Dataset
The dataset for this project is the [Appliances Energy Prediction data](https://archive.ics.uci.edu/ml/machine-learning-databases/00374/). The data set is at 10 min for about 4.5 months. The house temperature and humidity conditions were monitored with a ZigBee wireless sensor network. Each wireless node transmitted the temperature and humidity conditions around 3.3 min. Then, the wireless data was averaged for 10 minutes periods. The energy data was logged every 10 minutes with m-bus energy meters. Weather from the nearest airport weather station (Chievres Airport, Belgium) was downloaded from a public data set from Reliable Prognosis (rp5.ru), and merged together with the experimental data sets using the date and time column. 

Two random variables have been included in the data set for testing the regression models and to filter out non predictive attributes (parameters).

**Below is a preview of a couple columns I'll work with from the dataset:**

- **Date**: time year-month-day hour:minute:second
- **Appliances**: energy use in Wh
- **Lights**: energy use of light fixtures in the house in Wh
- **T1**: Temperature in kitchen area, in Celsius
- **RH_1**: Humidity in kitchen area, in %
- **T2**: Temperature in living room area, in Celsius
- **RH_2**: Humidity in living room area, in %
- **T3**: Temperature in laundry room area
- **RH_3**: Humidity in laundry room area, in %
- **T4**: Temperature in office room, in Celsius
- **RH_4**: Humidity in office room, in %
- **T5**: Temperature in bathroom, in Celsius
- **RH_5**: Humidity in bathroom, in %
- **T6**: Temperature outside the building (north side), in Celsius
- **RH_6**: Humidity outside the building (north side), in %
- **T7**: Temperature in ironing room , in Celsius
- **RH_7**: Humidity in ironing room, in %
- **T8**: Temperature in teenager room 2, in Celsius
- **RH_8**: Humidity in teenager room 2, in %
- **T9**: Temperature in parents room, in Celsius
- **RH_9**: Humidity in parents room, in %
- **To**: Temperature outside (from Chievres weather station), in Celsius
- **Pressure** (from Chievres weather station), in mm Hg
- **RH_out**: Humidity outside (from Chievres weather station), in %
- **Wind speed** (from Chievres weather station), in m/s
- **Visibility** (from Chievres weather station), in km
- **Tdewpoint** (from Chievres weather station), Â °C
- **rv1**, Random variable 1, nondimensional
- **rv2**, Random variable 2, nondimensional

**NOTE** : 
To answer some questions, you will need to normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set.


## Opening and Exploring the Data

Lets start by opening the dataset and then continue exploring the datset

In [1]:
#read dataset from github into a dataframe
import numpy as np
import pandas as pd
url = 'https://github.com/doyinsolamiolaoye/hamoye_internship/blob/master/energydata_complete.csv?raw=true'
df = pd.read_csv(url)

In [2]:
#number of rows and columns of the dataset
df.shape

(19735, 29)

In [3]:
#column labels
df.columns

Index(['date', 'Appliances', 'lights', 'T1', 'RH_1', 'T2', 'RH_2', 'T3',
       'RH_3', 'T4', 'RH_4', 'T5', 'RH_5', 'T6', 'RH_6', 'T7', 'RH_7', 'T8',
       'RH_8', 'T9', 'RH_9', 'T_out', 'Press_mm_hg', 'RH_out', 'Windspeed',
       'Visibility', 'Tdewpoint', 'rv1', 'rv2'],
      dtype='object')

In [4]:
#show the first 5 rows of the dataset
df.head()

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
0  2016-01-11 17:00:00          60      30  ...        5.3  13.275433  13.275433
1  2016-01-11 17:10:00          60      30  ...        5.2  18.606195  18.606195
2  2016-01-11 17:20:00          50      30  ...        5.1  28.642668  28.642668
3  2016-01-11 17:30:00          50      40  ...        5.0  45.410389  45.410389
4  2016-01-11 17:40:00          60      40  ...        4.9  10.084097  10.084097

[5 rows x 29 columns]

In [5]:
df.tail()

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
19730  2016-05-27 17:20:00         100       0  ...  13.333333  43.096812  43.096812
19731  2016-05-27 17:30:00          90       0  ...  13.300000  49.282940  49.282940
19732  2016-05-27 17:40:00         270      10  ...  13.266667  29.199117  29.199117
19733  2016-05-27 17:50:00         420      10  ...  13.233333   6.322784   6.322784
19734  2016-05-27 18:00:00         430      10  ...  13.200000  34.118851  34.118851

[5 rows x 29 columns]

In [6]:
df.describe()

Appliances        lights  ...           rv1           rv2
count  19735.000000  19735.000000  ...  19735.000000  19735.000000
mean      97.694958      3.801875  ...     24.988033     24.988033
std      102.524891      7.935988  ...     14.496634     14.496634
min       10.000000      0.000000  ...      0.005322      0.005322
25%       50.000000      0.000000  ...     12.497889     12.497889
50%       60.000000      0.000000  ...     24.897653     24.897653
75%      100.000000      0.000000  ...     37.583769     37.583769
max     1080.000000     70.000000  ...     49.996530     49.996530

[8 rows x 28 columns]

I will drop the columns that are not needed for this project:

In [7]:
df.drop(['lights','date'], axis=1,inplace=True)

## Solution to Quiz

In [8]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [9]:
#Firstly, we normalise our dataset to a common scale using the min max scaler
scaler = MinMaxScaler()
normalised_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)
normalised_df.head()

Appliances       T1      RH_1  ...  Tdewpoint       rv1       rv2
0    0.046729  0.32735  0.566187  ...   0.538462  0.265449  0.265449
1    0.046729  0.32735  0.541326  ...   0.533937  0.372083  0.372083
2    0.037383  0.32735  0.530502  ...   0.529412  0.572848  0.572848
3    0.037383  0.32735  0.524080  ...   0.524887  0.908261  0.908261
4    0.046729  0.32735  0.531419  ...   0.520362  0.201611  0.201611

[5 rows x 27 columns]

### Question 12 - From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius(x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two D.P?

Simple linear regression

In [10]:
#Split the Data into predictors (X) and target variables (y)
X = normalised_df[['T2']]
y = normalised_df[['T6']]

# splitting data into training and testing set
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

reg = LinearRegression()
reg.fit(X_train, y_train)

pred = reg.predict(X_test)

#calculate the r-squared value to 2dp
round(r2_score(y_test, pred), 2)

0.64

Multivariate linear regression

In [11]:
#Split the Data into predictors (X) and target variables (y)
X = normalised_df.drop(columns=['Appliances'])
y = normalised_df['Appliances']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

reg_multi = LinearRegression()
reg_multi.fit(X_train, y_train)

pred = reg_multi.predict(X_test)

### Question 13 - What is the Mean Absolute Error (in two decimal places)?

In [12]:
mae = mean_absolute_error(y_test, pred)
round(mae, 2)

0.05

### Question 14 - What is the Residual Sum of Squares (in two decimal places)?

In [13]:
rss = np.sum(np.square(y_test - pred))
round(rss, 2)

45.35

### Question 15 - What is the Root Mean Squares Error (in three decimal places)?

In [14]:
rmse = np.sqrt(mean_squared_error(y_test, pred))
round(rmse, 3)

0.088

### Question 16 - What is the Coefficient of Determination (in two decimal places)?


In [15]:
r2_score_2 = r2_score(y_test, pred)
round(r2_score_2, 2)

0.15

### Question 17 - Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [16]:
#comparing the effects of regularisation
def get_weights_df(model, feat, col_name):
    #this function returns the weight of every feature
    weights = pd.Series(model.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    #weights_df[col_name] = weights_df[col_name].round(3)
    return weights_df

In [17]:
linear_model_weights = get_weights_df(reg_multi, X_train, 'Linear_Model_Weight')

linear_model_weights.sort_values('Linear_Model_Weight', ascending  = False)

Features  Linear_Model_Weight
25         RH_1             0.553547
24           T3             0.290627
23           T6             0.236425
22    Tdewpoint             0.117758
21           T8             0.101995
20         RH_3             0.096048
19         RH_6             0.038049
18    Windspeed             0.029183
17           T4             0.028981
16         RH_4             0.026386
15         RH_5             0.016006
14   Visibility             0.012307
13           T7             0.010319
12  Press_mm_hg             0.006839
11          rv2             0.000770
10          rv1             0.000770
9            T1            -0.003281
8            T5            -0.015657
7          RH_9            -0.039800
6          RH_7            -0.044614
5        RH_out            -0.077671
4          RH_8            -0.157595
3            T9            -0.189941
2            T2            -0.236178
1         T_out            -0.321860
0          RH_2            -0.456698

### Question 18 - Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [18]:
# Define the Ridge model with an alpha value of 0.4
ridge_reg = Ridge(alpha = 0.4)
ridge_reg.fit(X_train, y_train)

# Make predictions
ridge_reg_pred = ridge_reg.predict(X_test)

# RMSE
ridge_rmse = np.sqrt(mean_squared_error(y_test, ridge_reg_pred))
round(ridge_rmse, 3)

0.088

Same value gotten with the linear regression model above: 0.088

### Question 19 - Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [19]:
# Define the Lasso model and call its fit method
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(X_train, y_train)

# Call the get_weights_df function and pass the appropriate parameters
lasso_model_weights = get_weights_df(lasso_reg, X_train, 'Lasso_Model_Weight')

np.count_nonzero(lasso_model_weights['Lasso_Model_Weight'])

4

### Question 20 - What is the new RMSE with the Lasso Regression (in 3 decimal places)?

In [20]:
lasso_reg_pred = lasso_reg.predict(X_test)

# RMSE
lasso_rmse = np.sqrt(mean_squared_error(y_test, lasso_reg_pred))
round(lasso_rmse, 3)

0.094

## Conclusion

It can be concluded from the analysis above that:
- There is no change in the Root Mean Square Error when both the linear Regressionn and ridge regression Models are used.
- The coefficient of determination of the multivariate linear regression is very low (0.15) and as such it can be concluded that the Model does not fit our data well; the accuracy of our best of fit line is low.